In [68]:
print(os.getpid())

129735


In [2]:
import os
import gc
import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import xgboost as xgb
import tables
import re

from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.metrics import make_scorer

# Limit threads for numerical libraries to manage CPU usage
os.environ["OPENBLAS_NUM_THREADS"] = "7"
os.environ["OMP_NUM_THREADS"] = "7"

base_dir = "/home/skovtun/Python_projects/Kaggle/Single_cell/"
data_dir = os.path.join(base_dir, "data")
random_state = 77

os.chdir(data_dir)

In [67]:
#Getting external file providing the mapping of human genes to their chromosome coordinates on 
#the GRCh38 genome to use for reducing number of columns for every target.
# RAW LINE:
# 1	havana	gene	11869	14409	.	+	.	gene_id "ENSG00000223972"; gene_version "5"; gene_name "DDX11L1"; gene_source "havana"; gene_biotype "transcribed_unprocessed_pseudogene";

gtf_path = "Homo_sapiens.GRCh38.98.gtf"

genes = []

with open(gtf_path) as f:
    for line in f:
        #skipping comments
        if line.startswith("#"):
            continue
        #splitting the line
        fields = line.strip().split("\t")
        if fields[2] != "gene":
            continue
        
        chrom = fields[0]
        start = int(fields[3])
        end = int(fields[4])
        strand = str(fields[6])
        attr = fields[8]
        match_id = re.search(r'gene_id "([^"]+)"',attr)
        match_name = re.search(r'gene_name "([^"]+)"',attr)
        if match_id:
            gene_id = match_id.group(1)
        else:
            continue
        if match_name:
            gene_name = match_name.group(1)
        else:
            gene_name = None
        
        genes.append([gene_id, gene_name,strand, chrom, start, end])

gene_df = pd.DataFrame(genes, columns=["gene_id", "gene_name", "strand", "chr", "start", "end"])
mapping = {str(i): f"chr{i}" for i in range(1,23)}
mapping['X'] = 'chrX'
mapping['Y'] = 'chrY'
gene_df['chr'] = gene_df['chr'].map(mapping).fillna(gene_df['chr'])
#gene_df['chr'].unique()

In [79]:
gene_df.shape

(60623, 5)

In [78]:
#Getting a list of genes and all coordinates, parsing ATAC Peaks
path = "train_multi_inputs.h5"
cols = pd.read_hdf(path, key="train_multi_inputs", start=0, stop=1).columns
path = "train_multi_targets.h5"
cols_t = pd.read_hdf(path, key = 'train_multi_targets', start = 0, stop=1).columns

#From the column names of the multi creating a dataframe with location name, start and end.
arr = pd.Series(cols.values)
r_p = r'([^:]+):([\d]+)-([\d]+)'
chr_ranges = arr.str.extract(r_p)
chr_ranges.columns = ['chr','start','end']
chr_ranges['start'] = chr_ranges['start'].astype(int)
chr_ranges['end'] = chr_ranges['end'].astype(int)

#Reducing gene_df by choosing onnly chr present in chr_ranges and only gene_id's present as targets.
multi_chr = list(chr_ranges['chr'].unique())
gene_df_multi = gene_df[gene_df['chr'].isin(multi_chr)]
gene_df_multi = gene_df_multi.set_index('gene_id')
missing = cols_t.difference(gene_df_multi.index)
gene_df_multi = gene_df_multi.loc[gene_df_multi.index.intersection(cols_t)]
gene_df_multi.shape

(23404, 5)

In [79]:
gene_df_multi.head()

,gene_name,strand,chr,start,end
gene_id,,,,,
ENSG00000238009,AL627309.1,-,chr1,89295,133723
ENSG00000241860,AL627309.5,-,chr1,141474,173862
ENSG00000241599,AL627309.4,+,chr1,160446,161525
ENSG00000229905,AL669831.2,+,chr1,760911,761989
ENSG00000237491,LINC01409,+,chr1,778747,810065


In [5]:
#calculating maximum amount of features for every target.
window = 2000

genes = gene_df_multi.loc[gene_df_multi.index.intersection(cols_t)].copy()
if genes['strand']
genes["left"]  = genes["start"] - window
genes["right"] = genes["end"]   + window
genes["gene_id"] = genes.index  # preserve gene_id as a column

chr_ranges = chr_ranges.rename(columns={
    "start": "start_peak",
    "end":   "end_peak"
})

target_cols = pd.DataFrame()
results = []
gene_id = []
start_peak = []
end_peak = []

for chr_name, chr_peaks in chr_ranges.groupby("chr"):
    sub_genes = genes[genes["chr"] == chr_name]
    if sub_genes.empty or chr_peaks.empty:
        continue

    merged = sub_genes[['gene_id', 'chr', 'left', 'right']].merge(chr_peaks[['chr', 'start_peak', 'end_peak']], on="chr")

    mask = (
        (merged["start_peak"] >= merged["left"]) &
        (merged["end_peak"]   <= merged["right"])
    )
    m = merged[mask]
    results.append(m)
        
target_cols = pd.concat(results)
target_cols['col'] = target_cols['chr'].astype(str)+":"+target_cols['start_peak'].astype(str)+"-"+target_cols['end_peak'].astype(str)

In [58]:
target_cols.head()

,gene_id,chr,left,right,start_peak,end_peak,col
0,ENSG00000277400,GL000194.1,51590,117018,114519,115365,GL000194.1:114519-115365
1,ENSG00000277400,GL000194.1,51590,117018,55758,56597,GL000194.1:55758-56597
2,ENSG00000277400,GL000194.1,51590,117018,58217,58957,GL000194.1:58217-58957
3,ENSG00000277400,GL000194.1,51590,117018,59535,60431,GL000194.1:59535-60431
4,ENSG00000274847,GL000194.1,51594,117055,114519,115365,GL000194.1:114519-115365


In [37]:
#Loading pre-computed features from Notebook 1

X_pca = np.load("X_csr_1000.npy", mmap_mode="r")
if X_pca.dtype != np.float32:
    X_pca = np.asarray(X_pca, dtype=np.float32)

n_cells = X_pca.shape[0]

Y_pca = np.memmap("Y_train_pca_300.f32", dtype="float32", mode="r", shape=(n_cells, 300))
Y_components = np.load("Y_ipca_components_300.npy")
Y_mean = np.load("Y_ipca_mean.npy")

print("X_pca:", X_pca.shape, X_pca.dtype, type(X_pca))
print("Y_pca:", Y_pca.shape, Y_pca.dtype, type(Y_pca))


X_pca: (105942, 1000) float32 <class 'numpy.memmap'>
Y_pca: (105942, 300) float32 <class 'numpy.memmap'>


In [49]:
# loading metatdata
metadata_old = pd.read_csv('metadata.csv', index_col = 'cell_id')
fix = pd.read_csv('metadata_cite_day_2_donor_27678.csv', index_col = 'cell_id')
metadata = pd.concat([metadata_old, fix], axis = 0)
values = np.load("train_multi_cell_ids.npy") #cell id_s alligned with X_pca
cell_names = np.char.decode(values, encoding = 'utf-8')
del metadata_old, fix
multi_metadata  = metadata.loc[cell_names,['day','donor']]
del metadata
metadata_aligned = multi_metadata.loc[cell_names]
del multi_metadata
assert np.array_equal(metadata_aligned.index.values, cell_names)
day = metadata_aligned['day'].to_numpy(dtype=np.float32)
day = day.reshape(-1, 1)


In [53]:
#Normalizing day so it would be comparable to PCA features
d0 = (day - 2)/8
h4 = np.maximum(0, (day - 4)/8)
h7 = np.maximum(0, (day - 7)/8)
D = np.column_stack([d0, h4, h7])
X = np.empty((X_pca.shape[0], X_pca.shape[1] + 3), dtype = np.float32)
X[:,:X_pca.shape[1]] = X_pca
X[:,-3:] = D

In [ ]:
gene_access_2000 = np.zeros((n_cells, G), dtype=np.float32)

for j, gene in enumerate(top_genes):
    peak_idx = target_cols.get(gene, [])
    if len(peak_idx) == 0:
        continue

    # sum promoter peaks per cell (works fast with CSR)
    v = X_peaks_csr[:, peak_idx].sum(axis=1)          # (n_cells, 1) matrix
    gene_access_2000[:, j] = np.asarray(v).ravel().astype(np.float32, copy=False)


In [56]:
X_pca.shape[0]

105942

In [59]:
gene_importance = np.sum(np.abs(Y_components), axis = 0)
ranked_genes = pd.Series(gene_importance, index = cols_t).sort_values(ascending= False)
genes_2000 = ranked_genes[:2000].to_list()
gene_access_2000 = np.empty((X_pca.shape[0], 2000), dtype = np.float32)

    
target_cols_1000 = target_cols[target_cols['gene_id'].isin(genes_1000.index)]
peaks_1000 = target_cols_1000['col'].unique()
with tables.open_file("train_multi_inputs.h5", "r") as f:
    peaks = f.get_node("/train_multi_inputs/axis0")[:]
peaks_d = np.char.decode(peaks, encoding = 'utf-8')
peak_to_id = {str(peak): i for i,peak in enumerate(peaks_d)}
peaks_1000_idx = sorted(set([peak_to_id[peak] for peak in peaks_1000]))

In [71]:
chr_ranges.head()

,chr,start_peak,end_peak
0,GL000194.1,114519,115365
1,GL000194.1,55758,56597
2,GL000194.1,58217,58957
3,GL000194.1,59535,60431
4,GL000195.1,119766,120427


In [74]:
gene_df.head()

,gene_id,gene_name,strand,chr,start,end
0,ENSG00000223972,DDX11L1,+,chr1,11869,14409
1,ENSG00000227232,WASH7P,-,chr1,14404,29570
2,ENSG00000278267,MIR6859-1,-,chr1,17369,17436
3,ENSG00000243485,MIR1302-2HG,+,chr1,29554,31109
4,ENSG00000284332,MIR1302-2,+,chr1,30366,30503


In [77]:
genes.head()

,gene_name,chr,start,end
gene_id,,,,
ENSG00000238009,AL627309.1,chr1,89295,133723
ENSG00000241860,AL627309.5,chr1,141474,173862
ENSG00000241599,AL627309.4,chr1,160446,161525
ENSG00000229905,AL669831.2,chr1,760911,761989
ENSG00000237491,LINC01409,chr1,778747,810065


In [82]:
# Calculating aggregated biologicaly important gene expression levels for top 2000 genes.
# TSS: For + strand: TSS = start
# For − strand: TSS = end
# Then a common promoter window is:
# TSS − 2000 bp to TSS + 500 bp

genes = gene_df_multi.loc[gene_df_multi.index.intersection(cols_t)].copy()
tss = np.where(genes["strand"]=='+', genes['start'], genes['end'])
genes["left"]  = genes["start"] - 2000
genes["right"] = genes["end"]   + 500
genes["gene_id"] = genes.index  # preserve gene_id as a column

chr_ranges = chr_ranges.rename(columns={
    "start": "start_peak",
    "end":   "end_peak"
})

target_cols = pd.DataFrame()
results = []
gene_id = []
start_peak = []
end_peak = []

for chr_name, chr_peaks in chr_ranges.groupby("chr"):
    sub_genes = genes[genes["chr"] == chr_name]
    if sub_genes.empty or chr_peaks.empty:
        continue

    merged = sub_genes[['gene_id', 'chr', 'left', 'right']].merge(chr_peaks[['chr', 'start_peak', 'end_peak']], on="chr")

    mask = (
        (merged["start_peak"] >= merged["left"]) &
        (merged["end_peak"]   <= merged["right"])
    )
    m = merged[mask]
    results.append(m)
        
target_cols_tss = pd.concat(results)
target_cols_tss['col'] = target_cols_tss['chr'].astype(str)+":"+target_cols_tss['start_peak'].astype(str)+"-"+target_cols_tss['end_peak'].astype(str)

In [84]:
target_cols_tss.head()

,gene_id,chr,left,right,start_peak,end_peak,col
0,ENSG00000277400,GL000194.1,51590,115518,114519,115365,GL000194.1:114519-115365
1,ENSG00000277400,GL000194.1,51590,115518,55758,56597,GL000194.1:55758-56597
2,ENSG00000277400,GL000194.1,51590,115518,58217,58957,GL000194.1:58217-58957
3,ENSG00000277400,GL000194.1,51590,115518,59535,60431,GL000194.1:59535-60431
4,ENSG00000274847,GL000194.1,51594,115555,114519,115365,GL000194.1:114519-115365


In [ ]:
import scipy
X_csr = scipy.sparse.load_npz('train_multi_cell.npz').tocsr()
n_cells = X_csr.shape[0]



In [123]:
gene_importance = np.sum(np.abs(Y_components), axis = 0)
ranked_genes = pd.Series(gene_importance, index = cols_t).sort_values(ascending= False)
genes_2000 = set(ranked_genes.head(2000).index)

with tables.open_file("train_multi_inputs.h5", "r") as f:
    peaks = f.get_node("/train_multi_inputs/axis0")[:]
peaks_d = np.char.decode(peaks, encoding = 'utf-8')
peak_to_id = {str(peak): i for i,peak in enumerate(peaks_d)}

tc = (target_cols_tss.loc[target_cols_tss['gene_id'].isin(genes_2000),["gene_id", "col"]]
    .copy()
    .drop_duplicates(ignore_index = True))

tc['X_peak_idx'] = tc['col'].map(peak_to_id).astype(np.int32)
#for each gene_id getting a list of X_peak_idx(column numbers in X_csr
gene_to_X_idx = tc.groupby("gene_id")["X_peak_idx"].agg(list)

genes_unique = gene_to_X_idx.index.to_numpy()
X_agg = np.zeros((X_crs.shape[0], len(genes_unique)), dtype=np.float32)

for i, gene in enumerate(genes_unique):
    index_list = gene_to_X_idx.loc[gene]
    if len(index_list) == 0:
        continue
    X_agg[:,i] = X_crs[:,index_list].mean(axis=1).ravel().astype(np.float32)
del X_csr

In [126]:
X_agg[:5,:20]

array([[0.        , 0.        , 0.        , 0.        , 0.        ,
        0.10496612, 0.        , 0.        , 0.        , 0.        ,
        0.24128264, 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.14006296, 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.16866402, 0.        , 0.        , 0.        , 0.        ,
        0.        , 1.0303524 , 0.        , 0.        , 0.        ,
        0.10919409, 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.09372857, 0.        ,
        0.        , 0.7493012 , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.77739304,
        0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.14433566, 0.        ,
        0.        , 0.        , 0.        , 0.        , 0.21262048,
        0.        , 0.        , 0.        , 0

In [129]:
X.shape

(105942, 1003)

In [132]:
print(type(X), type(X_agg))

<class 'numpy.ndarray'> <class 'numpy.ndarray'>


In [133]:
print("X:", X.shape, X.dtype, type(X))
print("X_agg:", X_agg.shape, X_agg.dtype, type(X_agg))
print("Y_pca:", Y_pca.shape, Y_pca.dtype, type(Y_pca))


X: (105942, 1003) float32 <class 'numpy.ndarray'>
X_agg: (105942, 1988) float32 <class 'numpy.ndarray'>
Y_pca: (105942, 300) float32 <class 'numpy.memmap'>


In [55]:
# gene_importance = np.sum(np.abs(Y_components), axis = 0)
# ranked_genes = pd.Series(gene_importance, index = cols_t).sort_values(ascending= False)
# genes_1000 = ranked_genes[:1000]
# target_cols_1000 = target_cols[target_cols['gene_id'].isin(genes_1000.index)]
# peaks_1000 = target_cols_1000['col'].unique()
# with tables.open_file("train_multi_inputs.h5", "r") as f:
#     peaks = f.get_node("/train_multi_inputs/axis0")[:]
# peaks_d = np.char.decode(peaks, encoding = 'utf-8')
# peak_to_id = {str(peak): i for i,peak in enumerate(peaks_d)}
# peaks_1000_idx = sorted(set([peak_to_id[peak] for peak in peaks_1000]))

In [8]:
# import scipy
# X_crs = scipy.sparse.load_npz('train_multi_cell.npz')
# Xgene_1000 = X_crs[:,peaks_1000_idx]
# print(f"Original shape: {X_crs.shape}")
# print(f"Sliced shape: {Xgene_1000.shape}")

Original shape: (105942, 228942)
Sliced shape: (105942, 15581)


In [9]:
# from scipy.sparse import hstack, csr_matrix
# X_pca_sparse = csr_matrix(X_csr_1000)
# X = hstack([X_pca_sparse,Xgene_1000])
# X.shape

(105942, 16581)

In [127]:
# --- Data Splitting and Feature Separation ---
# Random Train-Test Split (80/20)
with tables.open_file("train_multi_inputs.h5", "r") as f:
    values = f.get_node("/train_multi_inputs/axis1")[:]
cell_names = np.char.decode(values, encoding = 'utf-8')
cell_to_id = {str(name): i for i,name in enumerate(cell_names)}


random_state=77
rng = np.random.default_rng(random_state)
shuffled_names = rng.permutation(cell_names)

split_point = int(len(shuffled_names) * 0.8)
train_names = shuffled_names[:split_point]
test_names = shuffled_names[split_point:]

train_id = sorted([cell_to_id[name] for name in train_names])
test_id = sorted([cell_to_id[name] for name in test_names])


In [134]:
train_id.dtype

AttributeError: 'list' object has no attribute 'dtype'

In [100]:
# import numpy as np

# # 1. Length Check: Do we have the same number of rows?
# print(f"H5 Names Count: {len(cell_names)}")
# print(f"Meta Rows Count: {len(multi_metadata)}")
# assert len(cell_names) == len(multi_metadata), "❌ Mismatch in length!"

# # 2. Order Check: Is every single name in the exact same spot?
# # We compare the list of names from H5 vs the Index of the Dataframe
# are_aligned = np.array_equal(multi_metadata.index.values, cell_names)

# if are_aligned:
#     print("✅ VERIFIED: The metadata is perfectly aligned with the H5 file.")
# else:
#     print("❌ WARNING: The order is different!")
#     # Show the first mismatch if it failed
#     for i, (name_h5, name_meta) in enumerate(zip(cell_names, multi_metadata.index)):
#         if name_h5 != name_meta:
#             print(f"First mismatch at row {i}:")
#             print(f"  H5 says:   {name_h5}")
#             print(f"  Meta says: {name_meta}")
#             break

H5 Names Count: 105942
Meta Rows Count: 105942
✅ VERIFIED: The metadata is perfectly aligned with the H5 file.


In [11]:
# inputs_train, inputs_test = X[train_id,:], X[test_id,:]
# y_subset = Y_pca[:,:100]
# y_train, y_test = y_subset[train_id,:], y_subset[test_id,:]
# print(inputs_train.shape, inputs_test.shape, y_train.shape, y_test.shape)

(84753, 16581) (21189, 16581) (84753, 100) (21189, 100)


In [16]:
# meta_train, meta_test = multi_metadata.iloc[train_id], multi_metadata.iloc[test_id]
# day_col_train = meta_train['day'].values.reshape(-1,1)
# day_sparse_train = csr_matrix(day_col_train)
# x_train = hstack([inputs_train,day_sparse_train])

# day_col_test = meta_test['day'].values.reshape(-1,1)
# day_sparse_test = csr_matrix(day_col_test)
# x_test = hstack([inputs_test, day_sparse_test])


In [17]:
x_train.shape

(84753, 16582)

In [25]:
# # 1. Choosing a donor to_hold_out
# donors = meta_train['donor'].unique()
# rng = np.random.default_rng(seed=77) 
# val_donor = rng.choice(donors)

# print(f" Randomly Selected Holdout Donor: {val_donor}")

# # 2. Create Indices based on that random choice
# tr_idx = np.where(meta_train['donor']!= val_donor)[0]
# tr_mask = (meta_train['donor'] != val_donor).values

# va_idx = np.where(meta_train['donor'] == val_donor)[0]
# va_mask = (meta_train['donor'] == val_donor).values

 Randomly Selected Holdout Donor: 32606


In [135]:
import time, gc
import numpy as np
import xgboost as xgb

# PCs to evaluate (0-based): PCs 2,5,10,30
pc_panel = [1, 4, 9, 29]

cand_A = dict(
    max_depth=6,
    learning_rate=0.05,
    min_child_weight=20,
    colsample_bytree=0.30,
    subsample=0.80,
    reg_alpha=1.0,
    reg_lambda=1.0,
)

BASE_PARAMS = dict(
    objective="reg:squarederror",
    eval_metric="rmse",
    tree_method="hist",
    max_bin=128,
    n_jobs= 8,
    verbosity=0,
    seed=77,
)

paramsA = {**BASE_PARAMS, **cand_A}

Y = Y_pca
n_targets = Y.shape[1]

rng = np.random.default_rng(77)

train_id = np.array(train_id, dtype=np.int64)
test_id  = np.array(test_id, dtype=np.int64)

tr_sub = rng.choice(train_id, size=min(20000, train_id.size), replace=False)
va_sub = rng.choice(test_id,  size=min(12000,  test_id.size),  replace=False)

# Baseline: X only
X_tr_base = X[tr_sub]
X_va_base = X[va_sub]

# Augmented: [X, X_agg]
X_tr_aug = np.concatenate([X[tr_sub], X_agg[tr_sub]], axis=1)
X_va_aug = np.concatenate([X[va_sub], X_agg[va_sub]], axis=1)
Y_tr = Y[tr_sub]
Y_va = Y[va_sub]

print("\nTrain rows:", X_tr_base.shape[0], " Val rows:", X_va_base.shape[0])
print("Baseline cols:", X_tr_base.shape[1])
print("Augmented cols:", X_tr_aug.shape[1])

# -----------------------------
# METRIC: Pearson(y_true, y_pred) for each PC
# -----------------------------
def pearson_1d(a, b):
    a = a.astype(np.float32, copy=False)
    b = b.astype(np.float32, copy=False)
    a = a - a.mean()
    b = b - b.mean()
    denom = float(np.sqrt((a*a).sum()) * np.sqrt((b*b).sum()))
    return np.nan if denom == 0.0 else float((a*b).sum() / denom)


# -----------------------------
# RUN: baseline vs augmented
# -----------------------------
for name, X_tr, X_va in [("BASELINE_X", X_tr_base, X_va_base),
                         ("AUGMENTED_X_plus_Xagg", X_tr_aug, X_va_aug)]:

    print(f"\n=== {name} ===")
    t0 = time.perf_counter()

    # Build DMatrices once, reuse by swapping label
    j0 = pc_panel[0]
    dtrain = xgb.DMatrix(X_tr, label=Y_tr[:, j0])
    dvalid = xgb.DMatrix(X_va, label=Y_va[:, j0])

    per_pc = []
    for j in pc_panel:
        dtrain.set_float_info("label", Y_tr[:, j])
        dvalid.set_float_info("label", Y_va[:, j])

        booster = xgb.train(
            paramsA,
            dtrain,
            num_boost_round=800,
            evals=[(dvalid, "val")],
            early_stopping_rounds=30,
            verbose_eval=False
        )

        pred = booster.predict(dvalid, iteration_range=(0, booster.best_iteration + 1))
        pr = pearson_1d(Y_va[:, j], pred)

        per_pc.append((j, pr, int(booster.best_iteration), float(booster.best_score)))
        print(f"PC{j+1:>3}: pearson={pr:.5f} | best_iter={int(booster.best_iteration):>4} | rmse={float(booster.best_score):.4f}")

        del booster, pred
        gc.collect()

    mean_pr = float(np.nanmean([x[1] for x in per_pc]))
    total_s = time.perf_counter() - t0
    print(f"{name}: mean Pearson over PCs = {mean_pr:.5f} | time={total_s/60:.1f} min")

    del dtrain, dvalid
    gc.collect()


Train rows: 20000  Val rows: 12000
Baseline cols: 1003
Augmented cols: 2991

=== BASELINE_X ===
PC  2: pearson=0.89232 | best_iter= 570 | rmse=12.6970
PC  5: pearson=0.91477 | best_iter= 571 | rmse=5.9619
PC 10: pearson=0.74551 | best_iter= 352 | rmse=6.5859
PC 30: pearson=0.56268 | best_iter= 232 | rmse=3.3273
BASELINE_X: mean Pearson over PCs = 0.77882 | time=2.0 min

=== AUGMENTED_X_plus_Xagg ===
PC  2: pearson=0.89052 | best_iter= 436 | rmse=12.8145
PC  5: pearson=0.91343 | best_iter= 379 | rmse=6.0099
PC 10: pearson=0.74156 | best_iter= 356 | rmse=6.6322
PC 30: pearson=0.55926 | best_iter= 288 | rmse=3.3354
AUGMENTED_X_plus_Xagg: mean Pearson over PCs = 0.77619 | time=4.9 min


In [136]:
import numpy as np
import tables
from sklearn.linear_model import Ridge
import time

# -----------------------------
# CONFIG
# -----------------------------
SEED = 77
TRAIN_SUB = 20000
VAL_SUB   = 12000

N_PCS = 300
N_CORR_GENES = 2000          # residual correct only top genes
GENE_CHUNK = 2000            # for streaming pearson

H5_PATH = "train_multi_targets.h5"
COMP_PATH = "Y_ipca_components_300.npy"
MEAN_PATH = "Y_ipca_mean.npy"

# -----------------------------
# 0) Indices (use your split)
# -----------------------------
rng = np.random.default_rng(SEED)

train_id = np.array(train_id, dtype=np.int64)
test_id  = np.array(test_id,  dtype=np.int64)

tr_sub = rng.choice(train_id, size=min(TRAIN_SUB, train_id.size), replace=False)
va_sub = rng.choice(test_id,  size=min(VAL_SUB,  test_id.size),  replace=False)

# -----------------------------
# 1) Base model: X -> Y_pca (300)
# -----------------------------
# Use Ridge for speed + stability; you can swap to your XGB-per-PC later.
X_tr = X[tr_sub]     # (n_tr, 1003)
X_va = X[va_sub]     # (n_va, 1003)

Yp_tr = Y_pca[tr_sub, :N_PCS]   # memmap slice -> ndarray view/copy as needed
Yp_va = Y_pca[va_sub, :N_PCS]

t0 = time.time()
base = Ridge(alpha=1.0, random_state=SEED)  # simple baseline; tune alpha later if needed
base.fit(X_tr, Yp_tr)
Z_tr = base.predict(X_tr).astype(np.float32, copy=False)   # (n_tr, 300)
Z_va = base.predict(X_va).astype(np.float32, copy=False)   # (n_va, 300)
print(f"Base Ridge fit+pred: {(time.time()-t0)/60:.2f} min")

# -----------------------------
# 2) Load PCA decode params (PC -> genes)
# -----------------------------
components = np.load(COMP_PATH).astype(np.float32, copy=False)  # (300, n_genes)
mean_vec   = np.load(MEAN_PATH).astype(np.float32, copy=False)  # (n_genes,)

assert components.shape[0] == N_PCS
n_genes = components.shape[1]
print("n_genes:", n_genes)

# -----------------------------
# 3) Choose genes to correct (top by total loading magnitude)
# -----------------------------
gene_importance = np.sum(np.abs(components), axis=0)  # (n_genes,)
top_gene_idx = np.argsort(-gene_importance)[:N_CORR_GENES]  # (2000,)
top_gene_idx_sorted = np.sort(top_gene_idx)  # helpful for intersections

# -----------------------------
# 4) Build residual targets for those genes on TRAIN_SUB
#     residual = Y_true_gene - Y_pred_gene_base
# -----------------------------
# Decode base predictions only for the selected genes
# Y_pred_sub = Z_tr @ components[:, top_gene_idx] + mean[top_gene_idx]
C_sub = components[:, top_gene_idx]                       # (300, 2000)
mu_sub = mean_vec[top_gene_idx]                           # (2000,)

Ypred_tr_sub = (Z_tr @ C_sub + mu_sub).astype(np.float32, copy=False)  # (n_tr, 2000)

# Read true genes for those cells/genes from H5 (row slice, then col select)
with tables.open_file(H5_PATH, "r") as f:
    Y_node = f.get_node("/train_multi_targets/block0_values")  # (n_cells, n_genes)
    Ytrue_tr_full = Y_node[tr_sub, :]                          # (n_tr, n_genes)  (pulls n_tr x 24k)
    Ytrue_tr_sub = Ytrue_tr_full[:, top_gene_idx].astype(np.float32, copy=False)
    del Ytrue_tr_full

R_tr_sub = (Ytrue_tr_sub - Ypred_tr_sub).astype(np.float32, copy=False)  # (n_tr, 2000)

# -----------------------------
# 5) Residual model: X_agg -> residuals (for selected genes)
# -----------------------------
Xagg_tr = X_agg[tr_sub]   # (n_tr, 1988)
Xagg_va = X_agg[va_sub]   # (n_va, 1988)

t1 = time.time()
corr = Ridge(alpha=1.0, random_state=SEED)
corr.fit(Xagg_tr, R_tr_sub)
Rhat_va_sub = corr.predict(Xagg_va).astype(np.float32, copy=False)  # (n_va, 2000)
print(f"Residual Ridge fit+pred: {(time.time()-t1)/60:.2f} min")

# -----------------------------
# 6) Evaluate TRUE Kaggle-style mean cellwise Pearson in GENE SPACE on val rows
#     Streaming by gene chunks to avoid huge arrays.
#     Prediction = decode(Z_va) + residual_correction (only for top genes)
# -----------------------------
# Precompute mapping gene_idx -> column in residual matrix (0..1999), for fast lookup
# We'll use this to add correction only for genes in the current chunk.
pos_in_top = {g: k for k, g in enumerate(top_gene_idx)}

# Accumulate per-cell sums needed for Pearson across genes
n_va = Z_va.shape[0]
sum_y   = np.zeros(n_va, dtype=np.float64)
sum_y2  = np.zeros(n_va, dtype=np.float64)
sum_p   = np.zeros(n_va, dtype=np.float64)
sum_p2  = np.zeros(n_va, dtype=np.float64)
sum_yp  = np.zeros(n_va, dtype=np.float64)

with tables.open_file(H5_PATH, "r") as f:
    Y_node = f.get_node("/train_multi_targets/block0_values")

    for g0 in range(0, n_genes, GENE_CHUNK):
        g1 = min(n_genes, g0 + GENE_CHUNK)
        cols = slice(g0, g1)

        # True genes chunk for val rows
        Ytrue = Y_node[va_sub, cols].astype(np.float32, copy=False)      # (n_va, chunk)

        # Base prediction chunk via decoding
        # Ypred_base = Z_va @ components[:, cols] + mean[cols]
        C_chunk = components[:, g0:g1]                                   # (300, chunk)
        mu_chunk = mean_vec[g0:g1]                                       # (chunk,)
        Ypred = (Z_va @ C_chunk + mu_chunk).astype(np.float32, copy=False)

        # Add residual correction for genes in this chunk (only if gene is in top list)
        # Find which gene indices in [g0, g1) are in top_gene_idx
        # Do it by scanning the chunk indices (chunk <= 2000, cheap)
        if g1 > g0:
            # build list of (local_col_in_chunk, residual_col_in_Rhat) pairs
            add_pairs = []
            for gi in range(g0, g1):
                k = pos_in_top.get(gi)
                if k is not None:
                    add_pairs.append((gi - g0, k))

            if add_pairs:
                # apply additions column by column (pairs <= N_CORR_GENES, typically sparse per chunk)
                for local_j, k in add_pairs:
                    Ypred[:, local_j] += Rhat_va_sub[:, k]

        # Update Pearson sums per cell
        # Use float64 accumulators for numeric stability
        yt = Ytrue.astype(np.float64, copy=False)
        yp = Ypred.astype(np.float64, copy=False)

        sum_y  += yt.sum(axis=1)
        sum_y2 += (yt * yt).sum(axis=1)
        sum_p  += yp.sum(axis=1)
        sum_p2 += (yp * yp).sum(axis=1)
        sum_yp += (yt * yp).sum(axis=1)

# Compute per-cell Pearson across genes
G = float(n_genes)
num = sum_yp - (sum_y * sum_p) / G
den = np.sqrt((sum_y2 - (sum_y * sum_y) / G) * (sum_p2 - (sum_p * sum_p) / G))

cellwise = np.where(den > 0, num / den, 0.0)
mean_cellwise = float(np.mean(cellwise))

print("\nGene-space mean cellwise Pearson (val):", mean_cellwise)


Base Ridge fit+pred: 0.01 min
n_genes: 23418
Residual Ridge fit+pred: 0.02 min

Gene-space mean cellwise Pearson (val): 0.6554017252664152


In [137]:
pos_in_top = {g: k for k, g in enumerate(top_gene_idx)}

# Accumulate per-cell sums needed for Pearson across genes
n_va = Z_va.shape[0]
sum_y   = np.zeros(n_va, dtype=np.float64)
sum_y2  = np.zeros(n_va, dtype=np.float64)
sum_p   = np.zeros(n_va, dtype=np.float64)
sum_p2  = np.zeros(n_va, dtype=np.float64)
sum_yp  = np.zeros(n_va, dtype=np.float64)

with tables.open_file(H5_PATH, "r") as f:
    Y_node = f.get_node("/train_multi_targets/block0_values")

    for g0 in range(0, n_genes, GENE_CHUNK):
        g1 = min(n_genes, g0 + GENE_CHUNK)
        cols = slice(g0, g1)

        # True genes chunk for val rows
        Ytrue = Y_node[va_sub, cols].astype(np.float32, copy=False)      # (n_va, chunk)

        # Base prediction chunk via decoding
        # Ypred_base = Z_va @ components[:, cols] + mean[cols]
        C_chunk = components[:, g0:g1]                                   # (300, chunk)
        mu_chunk = mean_vec[g0:g1]                                       # (chunk,)
        Ypred = (Z_va @ C_chunk + mu_chunk).astype(np.float32, copy=False)

        # Add residual correction for genes in this chunk (only if gene is in top list)
        # Find which gene indices in [g0, g1) are in top_gene_idx
        # Do it by scanning the chunk indices (chunk <= 2000, cheap)
        if g1 > g0:
            # build list of (local_col_in_chunk, residual_col_in_Rhat) pairs
            add_pairs = []
            for gi in range(g0, g1):
                k = pos_in_top.get(gi)
                if k is not None:
                    add_pairs.append((gi - g0, k))

            # if add_pairs:
            #     # apply additions column by column (pairs <= N_CORR_GENES, typically sparse per chunk)
            #     for local_j, k in add_pairs:
            #         Ypred[:, local_j] += Rhat_va_sub[:, k]

        # Update Pearson sums per cell
        # Use float64 accumulators for numeric stability
        yt = Ytrue.astype(np.float64, copy=False)
        yp = Ypred.astype(np.float64, copy=False)

        sum_y  += yt.sum(axis=1)
        sum_y2 += (yt * yt).sum(axis=1)
        sum_p  += yp.sum(axis=1)
        sum_p2 += (yp * yp).sum(axis=1)
        sum_yp += (yt * yp).sum(axis=1)

# Compute per-cell Pearson across genes
G = float(n_genes)
num = sum_yp - (sum_y * sum_p) / G
den = np.sqrt((sum_y2 - (sum_y * sum_y) / G) * (sum_p2 - (sum_p * sum_p) / G))

cellwise = np.where(den > 0, num / den, 0.0)
mean_cellwise = float(np.mean(cellwise))

print("\nGene-space mean cellwise Pearson (val):", mean_cellwise)



Gene-space mean cellwise Pearson (val): 0.6664777166669354


Base Ridge fit+pred: 0.01 min
n_genes: 23418
Residual Ridge fit+pred: 0.02 min

Gene-space mean cellwise Pearson (val): 0.6554017252664152

In [ ]:
import numpy as np
import tables

# --------- inputs you already have ----------
components = np.load("Y_ipca_components_300.npy").astype(np.float32, copy=False)  # (300, n_genes)
mean_vec   = np.load("Y_ipca_mean.npy").astype(np.float32, copy=False)           # (n_genes,)

# Example: decode from Y_pca memmap you saved
Y_pca = np.memmap("Y_train_pca_300.f32", mode="r", dtype="float32")  # we'll reshape after reading n_cells

# Get n_cells, n_genes from the H5 (or store them elsewhere)
with tables.open_file("train_multi_targets.h5", "r") as f:
    Y_node = f.get_node("/train_multi_targets/block0_values")
    n_cells, n_genes = Y_node.shape

Y_pca = Y_pca.reshape(n_cells, components.shape[0])  # (n_cells, 300)

# --------- output memmap (full genes) ----------
Y_recon = np.memmap(
    "Y_recon_from_pca.f32",
    mode="w+",
    dtype="float32",
    shape=(n_cells, n_genes),
)

batch = 1024
for start in range(0, n_cells, batch):
    end = min(n_cells, start + batch)
    Zb = Y_pca[start:end, :]                              # (b, 300)
    Yb = Zb @ components + mean_vec                       # (b, n_genes)
    Y_recon[start:end, :] = Yb.astype(np.float32, copy=False)

Y_recon.flush()
print("Saved:", "Y_recon_from_pca.f32", "shape=", Y_recon.shape)


In [ ]:
Z_pred = np.load("Z_pred.npy").astype(np.float32, copy=False)  # (n_cells, 300)

# must match same n_cells, same 300
assert Z_pred.shape[1] == components.shape[0]
assert Z_pred.shape[0] == n_cells

Y_pred = np.memmap(
    "Y_pred_genes.f32",
    mode="w+",
    dtype="float32",
    shape=(n_cells, n_genes),
)

batch = 1024
for start in range(0, n_cells, batch):
    end = min(n_cells, start + batch)
    Y_pred[start:end, :] = (Z_pred[start:end, :] @ components + mean_vec).astype(np.float32, copy=False)

Y_pred.flush()
print("Saved:", "Y_pred_genes.f32", "shape=", Y_pred.shape)


In [ ]:
idx = np.array(va_sub, dtype=np.int64)  # your chosen cell indices
Z_sub = Y_pca[idx, :]                   # or Z_pred[idx, :]

Y_sub = (Z_sub @ components + mean_vec).astype(np.float32, copy=False)  # (len(idx), n_genes)
print("Decoded subset:", Y_sub.shape)
